File: server/crypto/ports.ts
import type { VaultId } from '../../lib/domain/identity';
import type {
  DekVersion,
  EncryptionNonce,
  SealedPayload,
  VaultDekMetadata,
} from './core';
import type { DataEncryptionKey } from './key-material';

export type KeyManagementPort = {
  generateDataKey(input: {
    readonly vaultId: VaultId;
    readonly dekVersion: DekVersion;
  }): Promise<{
    readonly metadata: VaultDekMetadata;
    readonly key: DataEncryptionKey;
  }>;
  unwrapDataKey(metadata: VaultDekMetadata): Promise<DataEncryptionKey>;
};

export type NonceGeneratorPort = {
  createNonce(): Promise<unknown>;
};

export type NonceReservationPort = {
  reserve(input: {
    readonly vaultId: VaultId;
    readonly dekVersion: DekVersion;
    readonly nonce: EncryptionNonce;
  }): Promise<boolean>;
};

export type Aes256GcmPort = {
  seal(input: {
    readonly key: DataEncryptionKey;
    readonly nonce: EncryptionNonce;
    readonly aad: string;
    readonly plaintext: Uint8Array;
  }): Promise<SealedPayload>;
  open(input: {
    readonly key: DataEncryptionKey;
    readonly nonce: EncryptionNonce;
    readonly aad: string;
    readonly sealedPayload: SealedPayload;
  }): Promise<Uint8Array>;
};

File: backend/internal/cryptocontent/ports.go
package cryptocontent

import (
	"context"

	"github.com/fukamu/notes/backend/internal/identity"
)

type KeyManagementPort interface {
	GenerateDataKey(context.Context, identity.VaultID, DEKVersion) (VaultDEKMetadata, *DataEncryptionKey, error)
	UnwrapDataKey(context.Context, VaultDEKMetadata) (*DataEncryptionKey, error)
}

type NonceGeneratorPort interface {
	CreateNonce(context.Context) ([]byte, error)
}

type NonceReservationPort interface {
	ReserveNonce(context.Context, identity.VaultID, DEKVersion, string) (bool, error)
}

type AES256GCMPort interface {
	Seal(*DataEncryptionKey, string, string, []byte) (string, error)
	Open(*DataEncryptionKey, string, string, string) ([]byte, error)
}
